scipy/sparse/sparsetools/csr.h

```diff
@@ -790,65 +815,121 @@ void csr_binop_csr_canonical(const I n_row, const I n_col,
+                        Cx[nnz] = result;
+                        nnz++;
+                    }
+                    A_pos++;
+                } else {
+                    //B_j < A_j
+                    T result = op(0, Bx[B_pos]);
+                        if(result != 0) {
+                        Cj[nnz] = B_j;
+                        Cx[nnz] = result;
+                        nnz++;
+                    }
+                    B_pos++;
+                }
+            }
+        //operations of addition and subtraction will
+        //be performed in this block.
+        else
+            //while not finished with either row.
+            while(A_pos < A_end && B_pos < B_end){
+                I A_j = Aj[A_pos];
+                I B_j = Bj[B_pos];
+
+                if(A_j == B_j){
+                    T result = op(Ax[A_pos], Bx[B_pos]);
+                    //removes the implicit zeros generated in
+                    //the binary operation.
+                    if(result != 0 || Ax[A_pos] == 0){
+                        Cj[nnz] = A_j;
+                        Cx[nnz] = result;
+                        nnz++;
+                    }
+                    A_pos++;
+                    B_pos++;
+                } else if (A_j < B_j) {
                     Cj[nnz] = A_j;
-                    Cx[nnz] = result;
+                    Cx[nnz] = Ax[A_pos];
                     nnz++;
+                    A_pos++;
+                } else {
+                    //B_j < A_j
+                    Cj[nnz] = B_j;
+                    Cx[nnz] = op(0, Bx[B_pos]);
+                    nnz++;
+                    B_pos++;
                 }
-                A_pos++;
-                B_pos++;
-            } else if (A_j < B_j) {
-                T result = op(Ax[A_pos],0);
-                if (result != 0){
-                    Cj[nnz] = A_j;
+            }
+
+        //tail
+        if(!addsub){
+            while(A_pos < A_end){
+                T result = op(Ax[A_pos], 0);
+                if(result != 0) {
+                    Cj[nnz] = Aj[A_pos];
                     Cx[nnz] = result;
                     nnz++;
                 }
                 A_pos++;
-            } else {
-                //B_j < A_j
-                T result = op(0,Bx[B_pos]);
-                if (result != 0){
-                    Cj[nnz] = B_j;
+            }
+            while(B_pos < B_end){
+                T result = op(0, Bx[B_pos]);
+                if(result != 0) {
+                    Cj[nnz] = Bj[B_pos];
                     Cx[nnz] = result;
                     nnz++;
                 }
                 B_pos++;
             }
-        }
-
-        //tail
-        while(A_pos < A_end){
-            T result = op(Ax[A_pos],0);
-            if (result != 0){
+        //tail for addition and subtraction
+       } else{
+            while(A_pos < A_end){
                 Cj[nnz] = Aj[A_pos];
-                Cx[nnz] = result;
+                Cx[nnz] = Ax[A_pos];
```

Comment:
`op(0, Bx[B_pos])` for Bx but just `Ax[A_pos]` for Ax because if binary operation is minus Bx has to be multiplied with negative sign but this is not the case with Ax .